### PR TITLE
Support Roslyn 3.11, 4.4, 4.6, and 4.8

### LINKED ISSUE #66: Akka.Analyzers may not be compatible with .NET Framework v4.7
Need more data, but wanted to create an issue for this - looks like this is an SDK / Tooling issue with Roslyn.

### PR DESCRIPTION
**Pull request by Aaronontheweb**
## Changes

close #66

## Checklist

For significant changes, please ensure that the following have been completed (delete if not relevant):

* [x] This change follows the [Akka.NET API Compatibility Guidelines](https://getakka.net/community/contributing/api-changes-compatibility.html).
* [x] This change follows the [Akka.NET Wire Compatibility Guidelines](https://getakka.net/community/contributing/wire-compatibility.html).
* [x] I have [reviewed my own pull request](https://getakka.net/community/contributing/index.html#review-your-own-pull-requests).
* [x] Design discussion issue #66

### CODE REVIEW

File: .github/workflows/pr_validation.yaml [hunk 1]
```diff
@@ -41,8 +41,16 @@ jobs:
       - name: "dotnet build"
         run: dotnet build -c Release
 
+      # .NET Framework tests can't run reliably on Linux, so we only do .NET 8
+
       - name: "dotnet test"
-        run: dotnet test -c Release
+        shell: bash
+        run: |
+          if [ "${{ matrix.os }}" = "ubuntu-latest" ]; then
```
Comment (Aaronontheweb) [inline .github/workflows/pr_validation.yaml:49]: Can't run .NET Framework tests reliably on Linux; worked around that here.

File: Directory.Build.props [hunk 2]
```diff
@@ -38,4 +44,35 @@
       </PropertyGroup>
     </When>
   </Choose>
+  
+  <Choose>
+    <When Condition="$(MSBuildProjectName.Contains('.Tests'))">
+      <ItemGroup>
+        <!-- Download packages referenced by ReferenceAssembliesHelper -->
+        <PackageDownload Include="Akka.Cluster.Sharding" Version="[1.5.15]"/>
```
Comment (Aaronontheweb) [inline Directory.Build.props:52]: Moved this to here so all of the individual `.Tests` projects could all download the same dependencies needed for running integration tests

File: Directory.Build.props [hunk 3]
```diff
@@ -21,6 +26,7 @@
     <LangVersion>latest</LangVersion>
     <ImplicitUsings>enable</ImplicitUsings>
     <Nullable>enable</Nullable>
+    <RoslynVersion>4.8.0</RoslynVersion>
```
Comment (Aaronontheweb) [inline Directory.Build.props:29]: The key driver of this Roslyn-versioning system is this custom property - this is used to specify the version of the Roslyn tooling that gets included inside each distributable.

File: Directory.Packages.props [hunk 4]
```diff
@@ -18,4 +18,8 @@
         <PackageVersion Include="FluentAssertions" Version="6.12.0"/>
         <PackageVersion Include="coverlet.collector" Version="6.0.1"/>
     </ItemGroup>
+
+    <ItemGroup>
+        <GlobalPackageReference Include="Microsoft.SourceLink.GitHub" Version="8.0.0" PrivateAssets="all" />
```
Comment (Aaronontheweb) [inline Directory.Packages.props:23]: Added SourceLink as a global package reference for all projects in the solution.

File: Directory.Build.props [hunk 5]
```diff
@@ -38,4 +44,35 @@
...
+  
+  <PropertyGroup Condition="$(MSBuildProjectName.StartsWith('Akka.Analyzers.Roslyn'))">
+    <AssemblyName>Akka.Analyzers</AssemblyName>
+  </PropertyGroup>
+  <PropertyGroup Condition="$(MSBuildProjectName.StartsWith('Akka.Analyzers.Fixes.Roslyn'))">
+    <AssemblyName>Akka.Analyzers.Fixes</AssemblyName>
+  </PropertyGroup>
+  <PropertyGroup Condition="$(MSBuildProjectName.Contains('Roslyn311'))">
+    <DefineConstants>$(DefineConstants);ROSLYN_3_11;ROSLYN_3_11_OR_GREATER</DefineConstants>
+    <RoslynVersion>3.11.0</RoslynVersion>
```
Comment (Aaronontheweb) [inline Directory.Build.props:65]: Defines the versions of Roslyn we use based on the naming convention of the projects

File: src/Akka.Analyzers.NuGet/Akka.Analyzers.nuspec [hunk 6]
```diff
@@ -22,6 +22,14 @@
 
         <file target="analyzers\dotnet\cs\" src="..\Akka.Analyzers\bin\$Configuration$\netstandard2.0\Akka.Analyzers.dll" />
         <file target="analyzers\dotnet\cs\" src="..\Akka.Analyzers.Fixes\bin\$Configuration$\netstandard2.0\Akka.Analyzers.Fixes.dll" />
-<!--        <file target="tools\" src="..\xunit.analyzers.fixes\tools\*.ps1" />-->
+
+        <file target="analyzers\dotnet\roslyn3.11\cs\" src="..\Akka.Analyzers.Fixes.Roslyn311\bin\$Configuration$\netstandard2.0\Akka.Analyzers.dll" />
```
Comment (Aaronontheweb) [inline src/Akka.Analyzers.NuGet/Akka.Analyzers.nuspec:26]: This is what makes the solution workable for end-users - this moves compiler-platform specific versions of the Roslyn analyzers + fixes into the directories expected by the .NET Compiler platform. So if the user is running version 4.3 of Roslyn, the system should fall back to using the 4.2 version specified in this directory.

File: src/Akka.Analyzers/Akka.Analyzers.csproj [hunk 7]
```diff
@@ -2,12 +2,9 @@
...
     <PropertyGroup>
         <TargetFramework>netstandard2.0</TargetFramework>
-        <ImplicitUsings>enable</ImplicitUsings>
-        <Nullable>enable</Nullable>
-        <IsPackable>false</IsPackable>
     </PropertyGroup>
 
     <ItemGroup>
-        <PackageReference Include="Microsoft.CodeAnalysis"/>
+        <PackageReference Include="Microsoft.CodeAnalysis" VersionOverride="[$(RoslynVersion)]"/>
```
Comment (Aaronontheweb) [inline src/Akka.Analyzers/Akka.Analyzers.csproj:8]: This is where we specify the version of Roslyn tooling that will get used in the solution. All of the `.Roslyn{X}` solutions update their `RoslynVersion` variables accordingly and reference the tooling accordingly.

File: src/Akka.Analyzers/Utility/Guard.cs [hunk 8]
```diff
@@ -9,7 +9,7 @@ namespace Akka.Analyzers;
 /// <summary>
 ///     INTERNAL API
 /// </summary>
-internal static class Guard
+public static class Guard
```
Comment (Aaronontheweb) [inline src/Akka.Analyzers/Utility/Guard.cs:12]: So I don't have to declare every compiler version as a Friend assembly.

**Review by Aaronontheweb — COMMENTED** [review]
Borrowed this technique from xunit.analyzers - should work to resolve our multi-targeting issues for end users. The real problem isn't the version of .NET being targeted, it's the version of the Visual Studio / .NET SDK tooling being used: https://learn.microsoft.com/en-us/visualstudio/extensibility/roslyn-version-support?view=vs-2022

**Review by Arkatufus — APPROVED** [review]
LGTM

**Pull request merged**